Codegen breaks for proto definitions with a streaming service named 'Stream'

## Bug Report


The codegen module imports `futures_core::Stream` directly, but when using tonic-build to generate a service for proto definitions with a service named `Stream` the imported trait conflicts with the generated trait.

### Version

tonic 0.4.1, tonic-build 0.4.1, prost 0.7.0
<!--
List the versions of all `tonic` crates you are using. The easiest way to get
this information is using `cargo-tree`.

`cargo install cargo-tree`
(see install here: https://github.com/sfackler/cargo-tree)

Then:

`cargo tree | grep tonic`
-->

### Platform

`Linux bsull-xps15-arch 5.11.2-arch1-1 #1 SMP PREEMPT Fri, 26 Feb 2021 18:26:41 +0000 x86_64 GNU/Linux`

<!---
Output of `uname -a` (UNIX), or version and 32 or 64-bit (Windows)
-->

### Crates

tonic, tonic-build

### Description

Codegen breaks for proto definitions with a streaming service named 'Stream'.

I tried this code:

```
❯ cat Cargo.toml
[package]
name = "tonic-repro"
version = "0.1.0"
authors = ["Ben Sully <ben@bsull.io>"]
edition = "2018"

[dependencies]
tonic = "0.4.1"
prost = "0.7.0"

[build-dependencies]
tonic-build = "0.4.1"

❯ cat proto/test.proto
syntax = "proto3";
package test;

message TestMessage {
  bool ok = 1;
}

service Stream {
  rpc RunStream(TestMessage) returns (stream TestMessage);
}

❯ cat build.rs
fn main() -> Result<(), Box<dyn std::error::Error>> {
    Ok(tonic_build::compile_protos("proto/test.proto")?)
}

❯ cat src/lib.rs
mod test {
    tonic::include_proto!("test");
}
```

This fails to compile with:

```
❯ cargo c
    Checking tonic-repro v0.1.0 (/home/ben/repos/rust/tonic-repro)
error[E0220]: associated type `Item` not found for `stream_server::Stream`
  --> /home/ben/repos/rust/tonic-repro/target/debug/build/tonic-repro-52881560b6922b38/out/test.rs:78:38
   |
78 |         type RunStreamStream: Stream<Item = Result<super::TestMessage, tonic::Status>>
   |                                      ^^^^ associated type `Item` not found

error: aborting due to previous error

For more information about this error, try `rustc --explain E0220`.
error: could not compile `tonic-repro`
```

I have a quick fix at https://github.com/hyperium/tonic/compare/master...sd2k:codegen-dont-export-stream, let me know if a PR would be accepted :)

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -18,6 +18,7 @@ members = [
     "tests/extern_path/uuid",
     "tests/ambiguous_methods",
     "tests/extern_path/my_application",
-    "tests/integration_tests"
+    "tests/integration_tests",
+    "tests/stream_conflict",
 ]
 

--- a/tests/stream_conflict/Cargo.toml
+++ b/tests/stream_conflict/Cargo.toml
@@ -1,0 +1,15 @@
+[package]
+name = "stream_conflict"
+version = "0.1.0"
+authors = ["Ben Sully <ben@bsull.io>"]
+publish = false
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+tonic = { path = "../../tonic" }
+prost = "0.7"
+
+[build-dependencies]
+tonic-build = { path = "../../tonic-build" }

--- a/tests/stream_conflict/build.rs
+++ b/tests/stream_conflict/build.rs
@@ -1,0 +1,3 @@
+fn main() {
+    tonic_build::compile_protos("proto/stream_conflict.proto").unwrap();
+}

--- a/tests/stream_conflict/proto/stream_conflict.proto
+++ b/tests/stream_conflict/proto/stream_conflict.proto
@@ -1,0 +1,10 @@
+syntax = "proto3";
+package StreamConflict;
+
+message Message {
+  bool ok = 1;
+}
+
+service Stream {
+  rpc RunStream(Message) returns (stream Message);
+}

--- a/tests/stream_conflict/src/main.rs
+++ b/tests/stream_conflict/src/main.rs
@@ -1,0 +1,7 @@
+mod stream_conflict {
+    tonic::include_proto!("stream_conflict");
+}
+
+fn main() {
+    println!("Hello, world!");
+}

--- a/tonic-build/src/server.rs
+++ b/tonic-build/src/server.rs
@@ -182,7 +182,7 @@ fn generate_trait_methods<T: Service>(
 
                 quote! {
                     #stream_doc
-                    type #stream: Stream<Item = Result<#res_message, tonic::Status>> + Send + Sync + 'static;
+                    type #stream: futures_core::Stream<Item = Result<#res_message, tonic::Status>> + Send + Sync + 'static;
 
                     #method_doc
                     async fn #name(&self, request: tonic::Request<#req_message>)
@@ -198,7 +198,7 @@ fn generate_trait_methods<T: Service>(
 
                 quote! {
                     #stream_doc
-                    type #stream: Stream<Item = Result<#res_message, tonic::Status>> + Send + Sync + 'static;
+                    type #stream: futures_core::Stream<Item = Result<#res_message, tonic::Status>> + Send + Sync + 'static;
 
                     #method_doc
                     async fn #name(&self, request: tonic::Request<tonic::Streaming<#req_message>>)

--- a/tonic/src/codegen.rs
+++ b/tonic/src/codegen.rs
@@ -1,7 +1,7 @@
 //! Codegen exports used by `tonic-build`.
 
 pub use async_trait::async_trait;
-pub use futures_core::Stream;
+pub use futures_core;
 pub use futures_util::future::{ok, poll_fn, Ready};
 
 pub use http_body::Body as HttpBody;

